### PR TITLE
feat: Add IO estimation based query mode decider

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Resolve https://github.com/risingwavelabs/risingwave/issues/16695
- A query should be executed in distributed only if its estimated IO is larger than a threshold (default 64M), to avoid a schedule cost for a small query.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by fuyufjh — COMMENTED** [review]
A first step toward CBO 😄 

I'd like to be cautious. The current state looks more predictable to me. Welcome for comments from more people.

File: src/frontend/src/optimizer/plan_visitor/execution_mode_decider.rs [hunk 1]
```diff
@@ -13,18 +13,31 @@
...
+}
 
 impl ExecutionModeDecider {
     /// If the plan should run in local mode, return true; otherwise, return false.
     pub fn run_in_local_mode(plan: PlanRef) -> bool {
-        let mut decider = ExecutionModeDecider {};
-        decider.visit(plan)
+        let mut decider = ExecutionModeDecider {
+            estimated_scan_io_bytes: 0,
+        };
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_visitor/execution_mode_decider.rs:33]: Better to make it a method like `Decider::new`.

File: src/frontend/src/optimizer/plan_visitor/execution_mode_decider.rs [hunk 2]
```diff
@@ -42,6 +55,22 @@ impl PlanVisitor for ExecutionModeDecider {
...
+            .base
+            .ctx()
+            .session_ctx()
+            .env()
+            .catalog_reader()
+            .read_guard();
+        if let Some(table_stat) = reader
+            .table_stats()
+            .table_stats
+            .get(&batch_seq_scan.core().table_catalog.id().table_id())
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_visitor/execution_mode_decider.rs:68]: nit: Is it possible that there are multiple `Scan` node for the same table? Should we deduplicate them when estimating the scan bytes considering block cache?

**Review by BugenZhao — COMMENTED** [review]
The implementation LGTM

**Pull request closed without merging**